File: flutter-boilerplate/lib/lib/billing/subscribe_flow.dart
/// BE-019: Stripe 3DS/SCA recovery for a first subscription.
///
/// The backend answers `subscribeToPlan` with `reason: authentication_required`
/// plus the first invoice's PaymentIntent `clientSecret` when the bank wants
/// the customer to authenticate. The flow is: run the next action on-device
/// (`Stripe.instance.handleNextAction`), then ask the backend to
/// `finalizeSubscription` — it re-reads the subscription from Stripe and
/// provisions the tier once it is active. Pure orchestration so it is
/// testable without the Stripe SDK.
class SubscribeDeclined implements Exception {
  final String reason;
  const SubscribeDeclined(this.reason);

  @override
  String toString() => 'SubscribeDeclined($reason)';
}

Future<Map<String, dynamic>> completeSubscribeWithAuthentication(
  Map<String, dynamic> initial, {
  required Future<void> Function(String clientSecret) confirm,
  required Future<Map<String, dynamic>> Function(String stripeSubscriptionId)
      finalize,
  int maxRounds = 2,
}) async {
  var result = initial;
  for (var round = 0; round < maxRounds; round++) {
    if (result['success'] == true) return result;
    final reason = result['reason'] as String?;
    final clientSecret = result['clientSecret'] as String?;
    final subscriptionId = result['stripeSubscriptionId'] as String?;
    if (reason != 'authentication_required' ||
        clientSecret == null ||
        subscriptionId == null) {
      throw SubscribeDeclined(reason ?? 'declined');
    }
    await confirm(clientSecret);
    result = await finalize(subscriptionId);
  }
  if (result['success'] == true) return result;
  throw const SubscribeDeclined('authentication_failed');
}

File: next-js-boilerplate/src/lib/billing/subscribe-flow.ts
import type { SubscribeResult } from "@/api/server/billing/stripe";

/**
 * BE-019: drive a first subscription through Stripe's 3DS/SCA step when the
 * backend answers `requiresAction`. Pure orchestration so it can be tested
 * without Stripe Elements: `confirm` is Stripe.js' confirmCardPayment
 * (resolves to an error message or null), `finalize` is the BFF finalize
 * call. Loops at most a couple of times in case Stripe asks again.
 */
export async function completeSubscribeWithAuthentication(
  initial: SubscribeResult,
  deps: {
    confirm: (clientSecret: string) => Promise<string | null>;
    finalize: (stripeSubscriptionId: string) => Promise<SubscribeResult>;
    authenticationFailedMessage: string;
  },
  maxRounds = 2,
): Promise<SubscribeResult> {
  let result = initial;
  for (let round = 0; round < maxRounds; round++) {
    if (!result.requiresAction) return result;
    if (!result.clientSecret || !result.stripeSubscriptionId) {
      throw new Error(deps.authenticationFailedMessage);
    }
    const confirmError = await deps.confirm(result.clientSecret);
    if (confirmError) throw new Error(confirmError);
    result = await deps.finalize(result.stripeSubscriptionId);
  }
  if (result.requiresAction) throw new Error(deps.authenticationFailedMessage);
  return result;
}

/** Map a BFF decline `key`/`reason` to the checkout copy, falling back to the generic message. */
export function declineMessage(
  err: unknown,
  t: {
    declinedInsufficientFunds: string;
    declinedCard: string;
    authenticationRequired: string;
    subscriptionFailed: string;
  },
): string {
  const exception = (err as { exception?: { key?: string; reason?: string } })
    ?.exception;
  const reason =
    exception?.reason ??
    (exception?.key?.startsWith("billing.errors.")
      ? exception.key.slice("billing.errors.".length)
      : undefined);
  switch (reason) {
    case "insufficient_funds":
    case "insufficientFunds":
      return t.declinedInsufficientFunds;
    case "declined":
      return t.declinedCard;
    case "authentication_required":
    case "authenticationRequired":
      return t.authenticationRequired;
    default:
      return (err as Error)?.message || t.subscriptionFailed;
  }
}
